### DISCUSSION TITLE
Is there any way to know if a syscall is blocking?

### DISCUSSION THREAD

**Discussion by davidzeng0** [axboe/liburing#1064] [Q&A] [answered]
I'm using shutdown(2) and socket(2) for older kernels that may not support the io_uring opcodes. I'm rather certain socket(2) does not block, but shutdown(2), with regards to the TCP protocol, seems like it might block because it does I/O?

Is there any place to read about these syscalls or is asking/reading the source the only option?

Reading the source is always a bit difficult because of the numerous virtual functions.

**Comment by YoSTEALTH** [comment]
> I'm rather certain socket(2) does not block

by default `socket()` will block. You need to set `SOCK_NONBLOCK` like `socket(AF_INET, SOCK_STREAM | SOCK_NONBLOCK)`

**Reply by davidzeng0** [reply]
Thats for `recv(2)` and `send(2)` and other socket related actions, no? Creating the socket itself shouldn't block

**Reply by YoSTEALTH** [reply]
I think it does block(pretty sure but still a guess), also closing blocks as well.

**Comment by YoSTEALTH** [comment]
Rather than making `io_uring` work with old kernel, I would recommend trying to find a way to use new kernel so you can use the latest and greatest `liburing` + `io_uring` features. It will be a lot less headache.

**Reply by davidzeng0** [reply]
I'm using kernel 6.6 but not everyone who's using my code is using a kernel that recent. Some 5.15, and something as old as 5.6 also works.

**Reply by YoSTEALTH** [reply]
Aww, makes sense. Use the new `io_uring_prep_socket()` for kernel that support it and use `socket()` for older kernel. Even if socket and such does block, there isn't much you can do about it... blocking happens for faction of a second. If user are really concerned with speed they would have to use the new kernel with your software.

**Comment by axboe** [comment, marked as answer]
Looking at the source is the only option... shutdown(2) may indeed block, which is why io_uring internally offloads it to io-wq to be safe.

**Reply by davidzeng0** [reply]
Got it, thank you :)
